Converged Tooltip Spec (#17261)
microsoft/fluentui: packages/react-tooltip/Spec.md

- GitHub Epic issue: [Tooltip Convergence #16735](https://github.com/microsoft/fluentui/issues/16735)

### Tooltips in v8/Fabric

v8 tooltips use a `TooltipHost` wrapped around the target element to provide tooltip functionality. This creates a `div` around the element that listens for mouse and focus events.

The `Tooltip` component renders as a `Callout`, and supports all `Callout` props.

```jsx
<TooltipHost
  content="This is the tooltip content"
  // This id is used on the tooltip itself, not the host
  // (so an element with this id only exists when the tooltip is shown)
  id={tooltipId}
  calloutProps={calloutProps}
  styles={hostStyles}
>
  <DefaultButton aria-describedby={tooltipId}>Hover over me</DefaultButton>
</TooltipHost>
```

#### Drawbacks

There are a few drawbacks with this approach to adding tooltips, which are outlined in [☂ Tooltip: open issues to resolve in converged approach #15102](https://github.com/microsoft/fluentui/issues/15102), and summarized below:

- The wrapper `div` created by `TooltipHost` can cause layout issues for the component. It also doesn't always result in proper positioning for the tooltip.
- The API is overly complex.

> - It's not possible to have a tooltip within a tooltip

Reviewer: Is tooltip nesting a requirement ?

Author: No it isn't. I'll remove that line.

There may be a requirement to have tooltips inside of popups, but that is a different scenario.

- There's no coordination between tooltips on a page. For example, moving the mouse between two elements with tooltips should cause the second tooltip to appear immediately without fading in/out.

### Tooltips in v0/Northstar

> v0 tooltips use a `trigger` property to render the tooltip's target component. However, unlike v8 it does not create a wrapper `div` around the target component, but instead adds listeners to the target component's props.

Reviewer 2: doesn't adding a wrapper div interfere with styling? Always when we had wrappers, it was causing confusion, maybe @miroslavstastny or @layershifter would be able to provide examples

Author: Right--the `div` added by v8 causes problems. As far as I could tell by looking at the code, v0 does _not_ add a wrapper `div`. And the plan here is to also not add any wrapper DOM elements.

```jsx
<Tooltip content="Example tooltip" trigger={<Button content="A button" />} />
```

# Sample Code

There are a few ways to add a tooltip to a component.

> ## A component with a `tooltip` slot

Reviewer: Having a tooltip slot will make things very clean. I do have one concern about what components should have tooltip slots and where ? 

We should consider the risk be for API consistency in the future and how we handle requests like 

> Can you add a tooltip slot to XYZ component cause our team needs it

😄

Author: The answer to that request is "good news, you can add it yourself!". As long as the component is built following our state hooks pattern, then all they need to do is add the `useTooltipSlot` hook.

For example, I've been testing a prototype implementation by creating an `AvatarWithTooltip` component:
```tsx
const AvatarWithTooltip = React.forwardRef((props: AvatarWithTooltipProps, ref: React.Ref<HTMLElement>) => {
  const state = useAvatar(props, ref, { tabIndex: 0 });
  useTooltipSlot(state);
  useAvatarStyles(state);
  return renderAvatar(state);
});
```

Author: That said, we'll need to make sure we add the `tooltip` prop to all of our core components that should support them.

Author: I suppose that wouldn't work well.. only one of them would show up but I'm not sure offhand which one. That said, it's equally as easy to write the wrong thing like this:
```tsx
<TooltipTrigger tooltip="from outer trigger">
  <TooltipTrigger tooltip="from inner trigger">
    <Button />
  </TooltipTrigger>
</TooltipTrigger>
```

It may be worthwhile to specify _which_ one shows up, since we can't really prevent it. (Probably the outermost one should take priority?)

I really like the idea of the `tooltip="..."` prop, which is effectively syntactic sugar for wrapping with a `<TooltipTrigger>`, but is much cleaner and more straightforward to understand.

Reviewer 3: Imo, `tooltip` is easier but more confusing. We're much more verbose with the component instead of the prop and we're also consistent with the rest of our library.
Also, we avoid confusion of what is part of the component itself or an addition, as well as not needing the cross dependency.

Author: We had a poll in the spec review meeting, and had a strong majority of people preferring having only `TooltipTrigger`, and not an additional `tooltip` prop.
* (6) - Only `TooltipTrigger`
* (1) - Add `tooltip` prop
* (2) - No vote

> Components can choose to expose a pseudo-slot `tooltip`, by inheriting props from the `WithTooltipSlot` interface and using the `useTooltipSlot` hook.

Reviewer: what is a `pseudo-slot` ? are we going to restrict the shorthand props that can be there ?

Author: Pseudo-slot in that it's not actually rendered by the component; instead it's rendered by the TooltipManager. I'll clarify in the spec.

While this looks and acts like a normal slot to users, the `Tooltip` is not actually rendered by the component, but rather passed to the `TooltipManager` to render. The `useTooltipSlot` hook is lightweight (small bundle size impact); it does not have a direct dependency on the `Tooltip` or `TooltipManager` classes.

For example, to add the `tooltip` slot to Button:

```typescript
// Button.types.ts:
export interface ButtonProps extends WithTooltipSlot /*...*/ {
  // ...
}

// useButton.ts:
export const useButton = (/*...*/) => {
  const state = mergeProps(/*...*/);

  // ...
  useTooltipSlot(state);
  // ...

  return state;
};
```

Example usage:

```jsx
<Button tooltip="Example tooltip" />
<Button tooltip={<>Custom <b>Tooltip</b> content!</>} />
<Button tooltip={{ children: 'Placed Tooltip', placement: 'right' }} />
<Button tooltip={{ as: MyTooltip, children: 'You can even implement your own tooltip' }} />
```

> ## On any element using a ref

Reviewer 2: don't we need third way as well - that would not require having a ref, but would just add the tooltip to any rendered component? Ideally however, API of tooltip and popup should be similar

Reviewer: Not all built components consumed from libraries (internal or OSS) are guaranteed to forward refs, so sometimes using a ref just might be possible

Author: I replaced useTooltipRef with a wrapper "TooltipTrigger" component. It won't add any DOM elements (i.e. no wrapper <div>), but it'll just attach listeners to its child component.

> ## On any element using a ref

Reviewer: If a consumer needed to use a `ref` for other purposes than just the tooltip would we they be required to do ref merging ?

Author: Yeah, they'd need to use `useMergedRefs`

To attach a tooltip to a component that doesn't have a `tooltip` slot, use the `useTooltipRef` hook. It takes the same shorthand props that the `tooltip` slot does. This is slightly less ergonomic than the tooltip slot, and requires using a ref, but works with any element.

```jsx

> <div ref={useTooltipRef('Example tooltip', { id: 'exampleTooltipId' })} aria-describedby="exampleTooltipId">

Reviewer 2: alternatively:

const tooltipProps = useTooltip('Example tooltip');
<div {...tooltipProps} />

Reviewer 2: NIT: to make tooltips accessible, the trigger needs to be focusable, so it would probably be better to change the examples to include focusable elements only:
`<button ref={useTooltipRe} ... />

Reviewer 2: one more thing to think through - how would the markup look like if the consumer needs to add tooltip, but also needs to get a ref to the element? Since here I think the useTooltipRef takes over the whole ref

Reviewer: > how would the markup look like if the consumer needs to add tooltip, but also needs to get a ref to the element

Resolving another comment I made in favour of a single thread. I have a similar concern, will we oblige consumers to merge refs ?

Author: The problem with `useTooltip` returning props is that it could overwrite some existing props instead of merging them. For example, it'll replace any existing `onFocus` handler.

Ling and I talked today and I'm now pursuing an approach like a `<TooltipTrigger>` component that wraps another React component. It'll add/merge `onFocus`, `aria-describedby`, etc. to the child component's props, without creating a wrapper `div`. In that way it'd work similarly to the v0 Tooltip, except it'd delegate to TooltipManager to handle actually showing the tooltip.

- _The **target** is the element that the tooltip is anchored to (and the arrow points to)._

_Almost always, these will both be the same element, but it is possible to specify them separately, so the tooltip can show up adjacent to a different element than the one that triggered it._

## @fluentui/react-tooltip-provider

The `react-tooltip-provider` is a lightweight package that allows any component to hook into tooltip functionality.

### TooltipProps

The `TooltipProps` interface is defined in `react-tooltip-provider` so that components can specify the details of the tooltip without taking a dependency on the larger `react-tooltip` package.

```typescript
export interface TooltipProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  /**
   * How to position the tooltip relative to the target element. This is a "best effort" placement,
   * but the tooltip may be flipped to the other side if there is not enough room.
   *
   * @defaultvalue bottom
   */
  placement?: TooltipPlacement;

  /**
   * Color variant with a subtle look
   */
  subtle?: boolean;

  /**
   * Do not render an arrow pointing to the target element
   */

>   noArrow?: boolean;

Author: v0 uses `pointing` for this (opposite of `noArrow`/`arrowless`). Which do people like the most between them?

>   noArrow?: boolean;

Reviewer 4: Oh yeah, I could totally see that, I guess it depends on what we want the default to be since I think we want the default value of these props to be false.

>   noArrow?: boolean;

Reviewer: are we going to allow custom styling on the arrow ? potentially would qualify that as a slot ?

Note that the `TooltipManager` has a slot for the visible tooltip on its State, but it's not a public prop. Instead, the tooltip is shown/hidden using the `TooltipManagerApi` imperative interface, which is defined in the `react-tooltip-provider` package below.

```typescript
import { TooltipManagerApi } from '@fluentui/react-tooltip-provider';

export interface TooltipManagerProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  /**
   * Ref to the imperative interface to show and hide tooltips.
   */
  componentRef?: React.Ref<TooltipManagerApi>;
}

export const tooltipManagerShorthandProps = ['tooltip'] as const;

export type TooltipManagerState = ComponentState<
  TooltipManagerProps & {
    /**
     * The Tooltip being rendered, if any.
     */
    tooltip?: ShorthandProps<TooltipProps>;
  }
>;
```

# Structure

## Public

```tsx
<TooltipProvider>
  <Button tooltip="Hello world" /> {/* for example... */}
</TooltipProvider>
```

## Internal

`TooltipProvider`:

```tsx
<internal__TooltipProviderContext.Provider value={{ current: undefined }}>
  <slots.root {...slotProps.root} />

>   <React.Suspense fallback={null}>

Reviewer: I have heard different things about supporting server side rendering so I do not know for sure if it is a requirement. But I think `Suspense` is not available for SSR: https://reactjs.org/docs/code-splitting.html#reactlazy

Author: Yeah. The lazy-loading aspect is not necessarily a hard requirement of this design.

The bigger reason for splitting out the react-tooltip-provider and react-tooltip packages is so that e.g. `Button` can have a `tooltip` prop, but not have a direct dependency on react-tooltip. That way, if someone wants Button but doesn't need tooltips, they don't pay the cost of the react-tooltip package.

    </div>
  </div>
</div>
```

# Migration

## Migration from v8

The converged API does not support many of the custom features of the v8 tooltip. We may need to revisit and add additional features to the converged Tooltip if needed.

- `Tooltip`
  - `calloutProps` => Not supported. The arrow's visibility can be controlled using `noArrow`.
  - `componentRef` => Not supported. The tooltip can't be manually invoked.
  - `content="..."` => Either a component's `tooltip="..."` prop or the Tooltip's `children` prop.
  - `delay` => `showDelay`
  - `directionalHint` => `placement`
    - `DirectionalHint.topLeftEdge` => `placement="top-start"`
    - `DirectionalHint.topCenter` => `placement="top"`
    - `DirectionalHint.topRightEdge` => `placement="top-end"`
    - `DirectionalHint.topAutoEdge` => Not supported
    - `DirectionalHint.bottomLeftEdge` => `placement="bottom-start"`
    - `DirectionalHint.bottomCenter` => `placement="bottom"`
    - `DirectionalHint.bottomRightEdge` => `placement="bottom-end"`
    - `DirectionalHint.bottomAutoEdge` => Not supported
    - `DirectionalHint.leftTopEdge` => `placement="left-start"`
    - `DirectionalHint.leftCenter` => `placement="left"`

>     - `DirectionalHint.leftBottomEdge` => `placement="left-end"`

Reviewer: I drew up a positioning table for v8 and v0 in the `Menu` spec: https://github.com/microsoft/fluentui/blob/master/packages/react-menu/Spec.md#positionalignment-hints

What do you think about omitting left/right so that syntax is independent of LTR/RTL ?

    - `DirectionalHint.rightBottomEdge` => `placement="right-end"`
  - `directionalHintForRTL` => Automatic based on whether the target element is RTL
  - `maxWidth` => `style={{ maxWidth: ... }}`
  - `onRenderContent` => Set `children` to a custom render function
- `TooltipHost`
  - `calloutProps` => Not supported
  - `closeDelay` => `hideDelay` on the Tooltip
  - `hostClassName` => N/A, not needed because there is no TooltipHost
  - `onTooltipToggle` => Not supported
  - `overflowMode` => `onlyIfTruncated`
    - `TooltipOverflowMode.self` => `onlyIfTruncated="true"`
    - `TooltipOverflowMode.parent` => Set `targetElement` to the parent element, and `onlyIfTruncated` to true

## Migration from v0

- `Tooltip`
  - `content="..."` => Either a component's `tooltip="..."` prop or the Tooltip's `children` prop.
  - `defaultOpen` => Not supported
  - `flipBoundary` => Not supported
  - `mountNode` => ???
  - `mouseLeaveDelay` => `hideDelay`
  - `offset` => Not supported
  - `onOpenChange` => Not supported
  - `open` => Not supported
  - `overflowBoundary` => Not supported
  - `pointing` => `!noArrow`
  - `popperRef` => Not supported

>   - `position` => `placement`

Reviewer: ```suggestion
  - `position` => `placement`
  - `align` => `placement`
```

v0 uses two props for position and alignment.... it can be hard to distinguish which controls what to be honest so having the single string could make sense, but regardless we should specify the replacement here

  - `positionFixed` => Not supported
  - `target` => `targetElement`
  - `trigger` => The component that has the `tooltip="..."` prop set, or the element that has the ref from `useTooltipRef` attached.

# Behaviors

- Visibility

  - There is only ever one tooltip visible at once.

>   - The tooltip shows 250ms after the trigger element receieves either mouse/pointer hover or keyboard focus.

Reviewer 4: ```suggestion
  - The tooltip, by default, shows 250ms after the trigger element receives either mouse/pointer hover or keyboard focus.
```

    - _Exception_: if another tooltip is currently visible, the new tooltip will immediately replace the old one without any delay.
    - _Exception_: if `onlyIfTruncated` is true, then the tooltip will only show if the target element's content is overflowing.

>   - The tooltip hides 250ms after losing mouse/pointer hover or keyboard focus (or immediately if another tooltip is shown).

Reviewer 4: ```suggestion
  - The tooltip, by default, hides 250ms after losing mouse/pointer hover or keyboard focus (or immediately if another tooltip is shown).
```

  - The tooltip hides immediately when the user presses Esc.

- Placement

  - The tooltip is placed relative to its target element, based on the `placement` property.
  - The placement will likely be handled by the PopperJS open source package. The discussion about using Popper is in [RFC: Integrate Popper into Fluent #17209](https://github.com/microsoft/fluentui/pull/17209)

- Focus

  - Content within the tooltip is not focusable, and can't be interacted with directly by keyboard or mouse.

- Screen readers
  - The tooltip is connected to the trigger element using `aria-describedby`, which should result in the screen reader reading the tooltip when shown.

# Accessibility

- ARIA design pattern: [Tooltip Widget](https://www.w3.org/TR/wai-aria-practices-1.2/#tooltip)

> - The Tooltip root element will have `role="tooltip"`

Reviewer: nit: Could we just add this to expected DOM output ? I am not in favour of specific a11y requirements for components because they should really be a requirement of delivering the component

Author: Yup I missed this in the DOM output; I'll add it.

> I am not in favour of specific a11y requirements for components because they should really be a requirement of delivering the component

I'm not sure I understand what you mean by this?

Reviewer: Ah yeah, I think that if we need to have accessible markup for DOM output of a component let's just spec that as expected DOM output of a component and rely on a specific `Accessibility` section as little as possible.

However I do think that this might be kind of opinionated, so as I said it's really just a nit

Author: I think @jurokapsiar added the Accessibility section to the spec template, so he might have an opinion on this.

I think it's worthwhile to list details like this explicitly, so devs writing specs don't forget about a11y :)

> - The trigger element (with the `tooltip="..."` prop) will have `aria-describedby` set to the tooltip element's ID. The ID will be generated automatically if not supplied.

Reviewer 2: tooltip can serve as a description (`aria-describedby`) or lablel (`aria-label` or `aria-labelledby`) - and both are valid use cases that we need to support. The tricky part is that the tooltip content element needs to be available in the DOM always, if you reference it using describedby or labelledby. if you use aria-label, the tooltip content element does not need to be in DOM as the text is direclty in the aria-label attribute.

Author: Thanks for pointing out the fact that the tooltip content needs to always be rendered. I'll need to rethink some of this design to accommodate that.

I'll also add the ability for a tooltip to be used either for `aria-describedby` or `aria-labelledby`.

> - The Tooltip itself can never receive focus.

Reviewer 2: also, tooltip can never contain focusable elements, it would be good to add the infromation here as well as to the future best practices section of Tooltip documentation.

export type TooltipState = ComponentState<
  TooltipProps & {
    /**
     * The arrow that points to the target element.
     */
    arrow?: ObjectShorthandProps<React.HTMLAttributes<HTMLElement> & React.RefAttributes<HTMLElement>>;
  },
  /* ShorthandProps: */ typeof tooltipShorthandProps[number],
  /* DefaultedProps: */ 'arrow' | 'placement'
>;
```

### TooltipProvider

`TooltipProvider` is responsible for providing the actual implementation of `TooltipManagerApi` and `Tooltip`. It uses a React context to that contains those when it is included in the tree. This context will also be built into `FluentContext`.

```typescript
export interface TooltipProviderProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  // TooltipProvider has no additional props
}

export const tooltipProviderShorthandProps = [] as const;

export type TooltipProviderState = ComponentState<TooltipProviderProps, typeof tooltipProviderShorthandProps[number]>;
```

### TooltipManager

`TooltipManager` handles showing and hiding tooltips, including positioning and tracking the visible tooltip. There is only one instance at the root of the app.

> 

Reviewer 5: This Codesandbox shows the scenario with multiple documents: https://codesandbox.io/embed/7hoso
(host application creates an iframe and then mounts a React portal into it)

I don't think that we should handle now "one tooltip per document", but portals should be mounted into a proper document. Please let me know if you will need any help with it 😉

```typescript
import { TooltipManagerApi } from '@fluentui/react-tooltip-hook';

export interface TooltipManagerProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  /**
   * Ref to the imperative interface to show and hide tooltips.
   */
  componentRef?: React.Ref<TooltipManagerApi>;
}

export const tooltipManagerShorthandProps = ['tooltip'] as const;

export type TooltipManagerState = ComponentState<
  TooltipManagerProps & {
    /**
     * The Tooltip being rendered, if any.
     */
    tooltip?: ShorthandProps<TooltipProps>;
  }
>;
```

# Structure

## Public

```tsx
<TooltipProvider>
  <Button tooltip="Example tooltip" />
  <TooltipTrigger tooltip="Element wrapped with TooltipTrigger">
    <a href="http://example.com" />
  </TooltipTrigger>
</TooltipProvider>
```

## DOM

In this example, the mouse is hovering over the first tooltip in the example above

```tsx
<body>
  <div {/* TooltipProvider */}>
    <button aria-describedby="tooltip-1" onPointerEnter={...} onPointerLeave={...} onFocus={...} onBlur={...} />
    <a href="http://example.com" aria-describedby="tooltip-2" onPointerEnter={...} onPointerLeave={...} onFocus={...} onBlur={...} />
  </div>

>   <div {/* Tooltip portal */}>

Reviewer 6: Is the portal exclusive for tooltips or will it be shared with `Popups`? Which component owns it? It should be `ThemeProvider` as it needs to inject a theme as CSS variables.

Author: It doesn't necessarily need to be shared by Popup. We do need to ensure that tooltips are always layered above popups; not sure how easy that is if they share the same portal?

I was thinking the portal would be owned by `TooltipProvider`; and the provider would need to be under a `ThemeProvider` for theming support.
